rclpy/src/rclpy/signal_handler.cpp

```diff
@@ -272,42 +260,27 @@ rclpy_register_sigint_guard_condition(PyObject * Py_UNUSED(self), PyObject * arg
   new_gcs[count_gcs + 1] = NULL;
 
   // Swap the lists and free the old
-  rcl_guard_condition_t ** old_gcs;
-  rcutils_atomic_exchange(&g_guard_conditions, old_gcs, new_gcs);
+  rcl_guard_condition_t ** old_gcs = g_guard_conditions.exchange(new_gcs);
   if (NULL != old_gcs) {
     allocator.deallocate(old_gcs, allocator.state);
   }
```

Comment:
I know this is preexistent code, but why are we using atomics here?
If this is for thread-safety, it doesn't look okay.
We're maybe alraedy taking a mutex in the python code, if that's the case this is okay.
If not we should be taking a global mutex here.

Reply:
>  We're maybe alraedy taking a mutex in the python code, if that's the case this is okay.

There is a global mutex in the Python code; it's holding onto the GIL. I think the atomics are to make it safe to be interrupted by the signal handler.

Reply:
> There is a global mutex in the Python code; it's holding onto the GIL.

Yes, but the signal handler might get executed without the GIL locked, and that might be an issue here (main thread executing the signal handler and another thread deallocating the array).
I think we should be using the python signal module instead of C signal handlers, which only flags the main thread to run the signal handler later, avoiding most of this issues.

I would left this comment unaddressed here, as it's unrelated to this PR and I want to avoid conflicts with https://github.com/ros2/rclpy/pull/830.

Reply:
> Yes, but the signal handler might get executed without the GIL locked, and that might be an issue here (main thread executing the signal handler and another thread deallocating the array)

Ah, I didn't realize other threads kept running when a signal is received.